controllers/numaresourcesoperator_controller.go

```diff
@@ -318,41 +320,54 @@ func (r *NUMAResourcesOperatorReconciler) syncNodeResourceTopologyAPI(ctx contex
 	return (updatedCount == len(objStates)), err
 }
 
-func (r *NUMAResourcesOperatorReconciler) syncMachineConfigs(ctx context.Context, instance *nropv1.NUMAResourcesOperator, trees []nodegroupv1.Tree) (bool, error) {
+func (r *NUMAResourcesOperatorReconciler) syncMachineConfigs(ctx context.Context, instance *nropv1.NUMAResourcesOperator, trees []nodegroupv1.Tree) (mcpWaitForUpdatedFunc, error) {
 	klog.V(4).InfoS("Machine Config Sync start", "trees", len(trees))
 	defer klog.V(4).Info("Machine Config Sync stop")
 
 	existing := rtestate.FromClient(ctx, r.Client, r.Platform, r.RTEManifests, instance, trees, r.Namespace)
 
 	var err error
-	var updatedCount int
+	var mcpUpdatedFunc mcpWaitForUpdatedFunc
 	objStates := existing.MachineConfigsState(r.RTEManifests)
-	for _, objState := range objStates {
-		if err2 := controllerutil.SetControllerReference(instance, objState.Desired, r.Scheme); err2 != nil {
-			err = errors.Wrapf(err2, "failed to set controller reference to %s %s", objState.Desired.GetNamespace(), objState.Desired.GetName())
-			break
+	// Since 4.18 we're using a built-in SELinux policy,
+	// so the MachineConfig which applies the custom policy is no longer necessary.
+	// In case of operator upgrade from 4.1X → 4.18, it's necessary to remove the old MachineConfig,
+	// unless an emergency annotation is provided which forces the operator to use custom policy
+	if !instance.IsCustomPolicyEnabled() {
+		for _, objState := range objStates {
+			if !objState.IsNotFoundError() {
```

Comment:
this looks fragile. We want to target only the `MachineConfig`. So it's probably better to add new functionalities to `pkg/objectstate/rte`

Reply:
I'm also thinking about `deleteUnusedMachineConfigs`. We can perhaps reuse that function - or even better get rid of it.

Reply:
...or not? https://github.com/openshift-kni/numaresources-operator/pull/1029

Reply:
What you did in #1029 is true, but we still need to remove the machine config explicitly when moving to built-in policy, because the CR remain in the cluster.
And as long as it persist on the cluster the machine-config won't go anywhere

Reply:
OK, let's try to remove them in the reconciliation loop rather than explicitely in the `deleteXXX` functions.

Reply:
> this looks fragile. We want to target only the MachineConfig. So it's probably better to add new functionalities to pkg/objectstate/rte

We do. you have the `objStates := existing.MachineConfigsState(r.RTEManifests)` that filters the MachineConfigs out of all the objects for you.

> OK, let's try to remove them in the reconciliation loop rather than explicitely in the deleteXXX functions.

This code is part of the reconciliation loop
it's called in `syncMachineConfigs`

Reply:
yes, I commented before to starting the effort in #1029 . It seems this of yours is the best approach on the table, and the removal of the `deleteXXX` functions should be tried (and preferably done!) on the side